### DISCUSSION TITLE
Pass callbacks to Vulkan API

### DISCUSSION THREAD

**Discussion by courgeon** [dotnet/Silk.NET#778] [Q&A] [answered]
Hello.

I'm following a tutorial in Vulkan C++ (I'm new to Vulkan) and I managed to get a Vulkan context open.
I'm now struggling with porting some part of the tutorial : 
They're showing how to bind a callback method to the validation layers: (C++)
```
VkDebugUtilsMessengerCreateInfoEXT createInfo{};
createInfo.sType =VK_STRUCTURE_TYPE_DEBUG_UTILS_MESSENGER_CREATE_INFO_EXT;
createInfo.messageSeverity = VK_DEBUG_UTILS_MESSAGE_SEVERITY_VERBOSE_BIT_EXT |
VK_DEBUG_UTILS_MESSAGE_SEVERITY_WARNING_BIT_EXT |
VK_DEBUG_UTILS_MESSAGE_SEVERITY_ERROR_BIT_EXT;
createInfo.messageType = VK_DEBUG_UTILS_MESSAGE_TYPE_GENERAL_BIT_EXT
| VK_DEBUG_UTILS_MESSAGE_TYPE_VALIDATION_BIT_EXT |
VK_DEBUG_UTILS_MESSAGE_TYPE_PERFORMANCE_BIT_EXT;
```

Which translate nicely to C#

```
  DebugUtilsMessengerCreateInfoEXT debuggerInfo = new DebugUtilsMessengerCreateInfoEXT();
  debuggerInfo.SType = StructureType.DebugUtilsMessengerCreateInfoExt;
  debuggerInfo.MessageSeverity = DebugUtilsMessageSeverityFlagsEXT.DebugUtilsMessageSeverityVerboseBitExt |
      DebugUtilsMessageSeverityFlagsEXT.DebugUtilsMessageSeverityWarningBitExt |
      DebugUtilsMessageSeverityFlagsEXT.DebugUtilsMessageSeverityErrorBitExt;
  debuggerInfo.MessageType = DebugUtilsMessageTypeFlagsEXT.DebugUtilsMessageTypeGeneralBitExt |
     DebugUtilsMessageTypeFlagsEXT.DebugUtilsMessageTypeValidationBitExt |
     DebugUtilsMessageTypeFlagsEXT.DebugUtilsMessageTypePerformanceBitExt;
```

but then, they bind a callback (C++)

```
createInfo.pfnUserCallback = debugCallback;
```

debugCallback being declared as a simple C++ method.

But in c#, the attribute is a **PfnDebugUtilsMessengerCallbackEXT**, and I can't figure out how I'm supposed to send a method through that.  There is a constructor that takes a " **delegate * unmanaged[Cdecl]< TYPES... >ptr**  "
How do a convert a static method to this ?

Anyone with some experience of this ?

Thanks for your help !

Matt

**Comment by Perksey** [comment, marked as answer]
Try
```cs
createInfo.PfnUserCallback = (DebugUtilsMessengerCallbackFunctionEXT) DebugCallback;
```

`DebugUtilsMessengerCallbackFunctionEXT` is a delegate which implicitly casts to the Pfn type.

**Reply by courgeon** [reply]
Quick answer, thanks :) 

The compiler says "No overload for "DebugCallback" matches delegate 'DebugUtilsMessengerCallbackFunctionEXT'
Where DebugCallback is a public static method.

**Reply by courgeon** [reply]
nevermind, the method declaration was mistyped. It does work

**Reply by courgeon** [reply]
Thanks you !
